src/NServiceBus.Metrics/RawData/RawDataReporter.cs

```diff
@@ -0,0 +1,126 @@
+            : this(dispatcher, destination, headers, buffer, outputWriter, DefaultFlushSize, DefaultMaxSpinningTime)
+        { }
+
+        public RawDataReporter(IDispatchMessages dispatcher, string destination, Dictionary<string, string> headers, RingBuffer buffer, WriteOutput outputWriter, int flushSize, TimeSpan maxSpinningTime)
+        {
+            this.buffer = buffer;
+            this.flushSize = flushSize;
+            this.maxSpinningTime = maxSpinningTime;
+            this.outputWriter = entries => outputWriter(entries, writer);
+            this.dispatcher = dispatcher;
+            this.destination = new UnicastAddressTag(destination);
+            this.headers = headers;
+
+            memoryStream = new MemoryStream();
+            writer = new BinaryWriter(memoryStream);
+            cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public void Start()
+        {
+            reporter = Task.Run(async () =>
+            {
+                while (cancellationTokenSource.IsCancellationRequested == false)
+                {
+                    var totalSpinningTime = TimeSpan.Zero;
+
+                    // spin till either MaxSpinningTime is reached OR items to consume are more than FlushSize
+                    while (totalSpinningTime < maxSpinningTime)
+                    {
+                        if (cancellationTokenSource.IsCancellationRequested)
+                        {
+                            break;
+                        }
+
+                        var itemsToConsume = buffer.RoughlyEstimateItemsToConsume();
+                        if (itemsToConsume >= flushSize)
+                        {
+                            break;
+                        }
+
+                        totalSpinningTime += singleSpinningTime;
+                        await Task.Delay(singleSpinningTime).ConfigureAwait(false);
+                    }
+
+                    await Consume().ConfigureAwait(false);
+                }
+
+                // flush data before ending
+                await Consume().ConfigureAwait(false);
+            });
+        }
+
+        async Task Consume()
+        {
+            var consumed = buffer.Consume(outputWriter);
+
+            if (consumed > 0)
+            {
+                writer.Flush();
+                var body = memoryStream.ToArray(); // if only transport operation allowed ArraySegment<byte>...
+
+                // clean stream
+                memoryStream.SetLength(0);
+
+                var message = new OutgoingMessage(Guid.NewGuid().ToString(), headers, body);
+                var operation = new TransportOperation(message, destination);
+                try
+                {
+                    await dispatcher.Dispatch(new TransportOperations(operation), transportTransaction, new ContextBag()).ConfigureAwait(false);
```

Comment:
In theory this context bag could be made static and reused since we are doing raw dispatches